src/cac/vsss.rs

```diff
@@ -1,11 +1,62 @@
 use std::ops::{Add, Mul};
 
-use ark_ec::PrimeGroup;
+use ark_ec::{PrimeGroup, scalar_mul::BatchMulPreprocessing};
+use ark_ff::BigInteger;
+use ark_ff::PrimeField;
 use ark_ff::{Field, One, UniformRand, Zero};
 use ark_secp256k1::{Fr, Projective};
 use rand::Rng;
 use serde::{Deserialize, Serialize};
 
+pub struct Secp256k1 {
+    pub generator: BatchMulPreprocessing<Projective>,
+}
+
+impl Default for Secp256k1 {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Secp256k1 {
+    pub fn new() -> Self {
+        // 181*176*3 is roughly the number of generator multiplications we do
+        Self {
+            generator: BatchMulPreprocessing::new(Projective::generator(), 181 * 176 * 3),
+        }
+    }
+
+    // Replacement of BatchMulPreprocessing::batch_mul, which (1) uses rayon parallization
+    // and (2) converts the result to an affine point instead of a projective point.
+    fn generator_batch_mul(&self, scalars: &[Fr]) -> Vec<Projective> {
+        scalars.iter().map(|e| self.windowed_mul(e)).collect()
+    }
+
+    // copied from ark-ec/src/scalar_mul/mod.rs because it's not public
+    fn windowed_mul(&self, scalar: &Fr) -> Projective {
```

Comment:
would be nice to make an upstream pr to make it public so we don't have to copy code

Reply:
Worth doing if we have a good argument for it!